<a href="https://colab.research.google.com/github/RaviduSenavirathna/Sinhala-Word-Recognizer/blob/main/main.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Confirm that all image files in the dataset located at `/content/dataset` have been successfully converted to PNG format.

## Iterate Subfolders

### Subtask:
Iterate through each subfolder within the main dataset directory.


**Reasoning**:
To iterate through the subfolders, I need to import the 'os' module, list all entries in the 'dataset_path', filter them to identify directories, and store their paths in a list as instructed.



In [ ]:
import os

# Get a list of all entries (files and directories) in the dataset_path directory
all_entries = os.listdir(dataset_path)

# Filter the list to include only directories and store their full paths
subfolders = []
for entry in all_entries:
    full_path = os.path.join(dataset_path, entry)
    if os.path.isdir(full_path):
        subfolders.append(full_path)

print(f"Found {len(subfolders)} subfolders in '{dataset_path}':")
# Print the first 5 subfolders to verify
for i, folder in enumerate(subfolders[:5]):
    print(f"  {i+1}. {folder}")

Found 314 subfolders in '/content/dataset':
  1. /content/dataset/පු
  2. /content/dataset/පො
  3. /content/dataset/බෙ
  4. /content/dataset/වෑ
  5. /content/dataset/ය්


## Process Images in Subfolder

### Subtask:
For each subfolder, identify all image files and convert them to PNG format, saving them in the same subfolder.


**Reasoning**:
To perform image processing operations like conversion and saving, I need to import the `Image` module from the `PIL` (Pillow) library. I will also import `os` if not already imported to handle file system operations like path manipulation and deletion.



In [ ]:
from PIL import Image
import os

# Initialize counters for verification
converted_files_count = 0
deleted_files_count = 0

# Define supported image extensions to convert
image_extensions = (".jpg", ".jpeg", ".gif", ".bmp", ".tiff", ".webp")

for subfolder_path in subfolders:
    print(f"Processing subfolder: {subfolder_path}")
    for entry in os.listdir(subfolder_path):
        full_file_path = os.path.join(subfolder_path, entry)

        if os.path.isfile(full_file_path) and entry.lower().endswith(image_extensions) and not entry.lower().endswith(".png"):
            try:
                # Open the image
                with Image.open(full_file_path) as img:
                    # Construct new PNG file path
                    base_name = os.path.splitext(entry)[0]
                    png_file_path = os.path.join(subfolder_path, f"{base_name}.png")

                    # Convert and save as PNG
                    img.save(png_file_path, "PNG")
                    converted_files_count += 1

                    # Delete the original file
                    os.remove(full_file_path)
                    deleted_files_count += 1
                    print(f"  Converted '{entry}' to '{os.path.basename(png_file_path)}' and deleted original.")
            except Exception as e:
                print(f"  Error processing '{entry}' in '{subfolder_path}': {e}")

print(f"\nSummary: {converted_files_count} files converted to PNG and {deleted_files_count} original files deleted.")

Processing subfolder: /content/dataset/පු
  Converted 'පු.jpg' to 'පු.png' and deleted original.
Processing subfolder: /content/dataset/පො
  Converted 'පො.jpg' to 'පො.png' and deleted original.
Processing subfolder: /content/dataset/බෙ
  Converted 'බෙ.jpg' to 'බෙ.png' and deleted original.
Processing subfolder: /content/dataset/වෑ
  Converted 'වෑ.jpg' to 'වෑ.png' and deleted original.
Processing subfolder: /content/dataset/ය්
  Error processing 'ය්.jpg' in '/content/dataset/ය්': cannot identify image file '/content/dataset/ය්/ය්.jpg'
Processing subfolder: /content/dataset/ආ
  Converted 'ආ.jpg' to 'ආ.png' and deleted original.
Processing subfolder: /content/dataset/ගූ
  Converted 'ගූ.jpg' to 'ගූ.png' and deleted original.
Processing subfolder: /content/dataset/දී
  Converted 'දී.jpg' to 'දී.png' and deleted original.
Processing subfolder: /content/dataset/ණා
  Error processing 'ණා.jpg' in '/content/dataset/ණා': cannot identify image file '/content/dataset/ණා/ණා.jpg'
Processing subfolder

**Reasoning**:
The previous step attempted to convert all image files to PNG format. However, there were errors reported for some files. To confirm that all image files have been successfully converted to PNG format and no original non-PNG files remain, I need to iterate through the subfolders again and check for the presence of any non-PNG image files.



In [ ]:
remaining_non_png_files = []
for subfolder_path in subfolders:
    for entry in os.listdir(subfolder_path):
        full_file_path = os.path.join(subfolder_path, entry)
        if os.path.isfile(full_file_path) and entry.lower().endswith(image_extensions) and not entry.lower().endswith(".png"):
            remaining_non_png_files.append(full_file_path)

if remaining_non_png_files:
    print("WARNING: The following non-PNG image files still exist:")
    for f in remaining_non_png_files:
        print(f"  - {f}")
    print(f"Total {len(remaining_non_png_files)} non-PNG files remaining.")
else:
    print("All identifiable image files have been successfully converted to PNG format.")

# Also check if any .png files were not created where an error occurred
# This would require comparing the original list of image files to the current list of png files.
# Given the current error handling, files that failed conversion would still be present.
# The above check for remaining non-PNG files addresses this verification directly.

  - /content/dataset/ය්/ය්.jpg
  - /content/dataset/ණා/ණා.jpg
  - /content/dataset/ර්/ර්.jpg
  - /content/dataset/ම්/ම්.jpg
  - /content/dataset/තා/තා.jpg
  - /content/dataset/ලා/ලා.jpg
  - /content/dataset/ටා/ටා.jpg
  - /content/dataset/බ්/බ්.jpg
  - /content/dataset/ට්/ට්.jpg
  - /content/dataset/ල්/ල්.jpg
  - /content/dataset/රා/රා.jpg
  - /content/dataset/ක්/ක්.jpg
  - /content/dataset/ඩ්/ඩ්.jpg
  - /content/dataset/ඩා/ඩා.jpg
  - /content/dataset/හා/හා.jpg
  - /content/dataset/ත්/ත්.jpg
  - /content/dataset/පා/පා.jpg
  - /content/dataset/චා/චා.jpg
  - /content/dataset/කා/කා.jpg
  - /content/dataset/ච්/ච්.jpg
  - /content/dataset/මා/මා.jpg
  - /content/dataset/සා/සා.jpg
  - /content/dataset/ණ්/ණ්.jpg
  - /content/dataset/ඬා/ඬා.jpg
  - /content/dataset/ව්/ව්.jpg
  - /content/dataset/ඳා/ඳා.jpg
  - /content/dataset/ද්/ද්.jpg
  - /content/dataset/ඟා/ඟා.jpg
  - /content/dataset/දා/දා.jpg
  - /content/dataset/ඹා/ඹා.jpg
  - /content/dataset/නා/නා.jpg
  - /content/dataset/ස්/ස්.jpg
  - /con

## Final Task

### Subtask:
Confirm that all image files in the dataset have been successfully converted to PNG format.


## Summary:

### Q&A
No, not all image files in the dataset have been successfully converted to PNG format.

### Data Analysis Key Findings
*   The dataset contains 314 subfolders.
*   A total of 268 image files were successfully converted from their original format to PNG and their original files were deleted.
*   During the conversion process, 46 image files encountered errors such as "cannot identify image file", indicating potential corruption or unsupported formats not covered by the conversion logic.
*   After the conversion attempt, 46 non-PNG image files remained in the dataset, corresponding to those that failed conversion.

### Insights or Next Steps
*   Investigate the 46 image files that failed conversion to determine the root cause (e.g., file corruption, unusual formats) and implement robust error handling or alternative conversion methods for these specific cases.
*   Consider adding more comprehensive logging during the conversion process to capture specific errors and file paths for easier debugging and to identify patterns in failed conversions.


# Task
Inspect the failed image files in the `remaining_non_png_files` list to determine their actual file types, sizes, and potential reasons for conversion failure, and then summarize these findings.

## Inspect Failed Image Files

### Subtask:
Iterate through the list of `remaining_non_png_files` and for each file, try to determine its actual file type, size, and any other relevant properties that might explain the conversion failure. This could involve using libraries like `filetype` or attempting to open them with more detailed error capturing.


**Reasoning**:
To investigate the failed image files, I will iterate through the `remaining_non_png_files` list, record their size, and attempt to open each file with PIL to capture its format or any encountered error, storing these details for analysis.



In [ ]:
import os
from PIL import Image, UnidentifiedImageError

failed_file_details = []

print(f"Analyzing {len(remaining_non_png_files)} remaining non-PNG files...")

for file_path in remaining_non_png_files:
    file_info = {"path": file_path, "size_bytes": None, "format": None, "error": None}

    try:
        # Get file size
        file_info["size_bytes"] = os.path.getsize(file_path)

        # Attempt to open with PIL to determine format or catch specific errors
        with Image.open(file_path) as img:
            file_info["format"] = img.format
    except UnidentifiedImageError as e:
        file_info["error"] = f"UnidentifiedImageError: {e}"
    except FileNotFoundError:
        file_info["error"] = "FileNotFoundError: File not found."
    except Exception as e:
        file_info["error"] = f"General Error: {e}"

    failed_file_details.append(file_info)

print("\n--- Failed File Details ---")
for detail in failed_file_details:
    print(f"Path: {detail['path']}")
    print(f"  Size: {detail['size_bytes']} bytes")
    if detail["format"]:
        print(f"  Identified Format: {detail['format']}")
    if detail["error"]:
        print(f"  Error: {detail['error']}")
    print("-------------------------")


Analyzing 46 remaining non-PNG files...

--- Failed File Details ---
Path: /content/dataset/ය්/ය්.jpg
  Size: 239729 bytes
  Error: UnidentifiedImageError: cannot identify image file '/content/dataset/ය්/ය්.jpg'
-------------------------
Path: /content/dataset/ණා/ණා.jpg
  Size: 76890 bytes
  Error: UnidentifiedImageError: cannot identify image file '/content/dataset/ණා/ණා.jpg'
-------------------------
Path: /content/dataset/ර්/ර්.jpg
  Size: 240866 bytes
  Error: UnidentifiedImageError: cannot identify image file '/content/dataset/ර්/ර්.jpg'
-------------------------
Path: /content/dataset/ම්/ම්.jpg
  Size: 210819 bytes
  Error: UnidentifiedImageError: cannot identify image file '/content/dataset/ම්/ම්.jpg'
-------------------------
Path: /content/dataset/තා/තා.jpg
  Size: 85473 bytes
  Error: UnidentifiedImageError: cannot identify image file '/content/dataset/තා/තා.jpg'
-------------------------
Path: /content/dataset/ලා/ලා.jpg
  Size: 79087 bytes
  Error: UnidentifiedImageError: ca

## Summary of Failed Image File Inspection

### Analysis Key Findings:
*   **Total Failed Files:** 46 image files, all with a `.jpg` extension, failed to convert to PNG format.
*   **Common Error:** For every single failed file, the `PIL.UnidentifiedImageError: cannot identify image file` was raised. This error indicates that the Pillow library (PIL) was unable to recognize the internal structure or headers of these files as a valid image format, despite their `.jpg` extension.
*   **File Sizes:** The sizes of these failed files vary significantly, ranging from approximately 65 KB to 332 KB. This suggests that the issue is not uniformly related to file size (e.g., all being empty or very small corrupted files).
*   **Potential Reasons for Failure:**
    *   **File Corruption:** The files might be corrupted and thus unreadable by PIL.
    *   **Incorrect Extension:** The files might not actually be JPEG images, but have been incorrectly named with a `.jpg` extension. They could be entirely different file types or invalid data.
    *   **Unsupported Format Variations:** While PIL generally supports common image formats, there might be obscure or malformed variations of JPEG that it cannot parse.
    *   **Incomplete Downloads/Transfers:** The files might be incomplete or truncated due to issues during their initial creation or transfer.

### Conclusion:
The `UnidentifiedImageError` across all 46 failed `.jpg` files strongly suggests that these files are either corrupted or are not valid JPEG images as expected. Further investigation into the actual content of these files (e.g., using a file signature analysis tool or a hex editor) would be needed to definitively determine their true nature. For now, they cannot be processed as images by the current conversion method.

## Final Task

### Subtask:
Provide a final confirmation on the status of image file conversions in the dataset, incorporating insights from the analysis of failed files.


## Summary:

### Q&A
The status of image file conversions is that 46 files, all with a `.jpg` extension, failed to convert to PNG format. The inspection of these failed files revealed that they could not be identified as valid image files by the PIL library. Their sizes vary from approximately 65 KB to 332 KB.

### Data Analysis Key Findings
*   A total of 46 image files, all with a `.jpg` extension, failed to convert to PNG format.
*   For every single failed file, the `PIL.UnidentifiedImageError: cannot identify image file` was raised, indicating that the Pillow library could not recognize their internal structure as a valid image format.
*   The sizes of these failed files varied significantly, ranging from approximately 65 KB to 332 KB, suggesting the issue is not uniformly related to file size.
*   Potential reasons for failure include file corruption, incorrect extensions masking non-image data, unsupported format variations, or incomplete downloads/transfers.

### Insights or Next Steps
*   The consistent `UnidentifiedImageError` across all failed files strongly suggests they are either corrupted or not valid JPEG images, despite their `.jpg` extension.
*   Further investigation using file signature analysis tools or a hex editor is recommended to definitively determine the true nature and content of these 46 problematic files.
